src/AWS/Lambda/Events/Kafka.hs

```diff
@@ -0,0 +1,249 @@
+-- for a sample payload.
+data KafkaEvent = KafkaEvent {
+  eventSource      :: !EventSource,
+  -- | Only present when using the "Amazon MSK" event source mapping.
+  eventSourceArn   :: !(Maybe Text),
+  bootstrapServers :: !(NonEmpty Text),
+  -- | The map's keys look like @"${topicName}-${partitionNumber}"@
+  records          :: !(Map Text [Record])
+} deriving (Eq, Show, Generic)
+
+instance FromJSON KafkaEvent where
+  parseJSON = withObject "KafkaEvent" $ \o ->
+    let parseBootstrapServers
+          = maybe (fail "bootstrapServers: empty string") pure
+          . NE.nonEmpty
+          . T.splitOn ","
+    in KafkaEvent
+         <$> o .: "eventSource"
+         <*> o .:! "eventSourceArn"
+         <*> (o .: "bootstrapServers" >>= parseBootstrapServers)
+         <*> o .: "records"
+
+instance ToJSON KafkaEvent where
+  toJSON e = object $
+    [ "eventSource" .= eventSource e
+    , "bootstrapServers" .= T.intercalate "," (NE.toList (bootstrapServers e))
+    , "records" .= records e
+    ] <> maybeToList (("eventSourceArn" .=) <$> eventSourceArn e)
+
+data EventSource
+  = AwsKafka -- ^ @"aws:kafka"@
+  | SelfManagedKafka -- ^ @"SelfManagedKafka"@
+  deriving (Eq, Show, Bounded, Enum, Ord, Generic)
+
+instance FromJSON EventSource where
+  parseJSON = withText "EventSource" $ \case
+    "aws:kafka" -> pure AwsKafka
+    "SelfManagedKafka" -> pure SelfManagedKafka
+    t -> fail $ "unrecognised EventSource: \"" <> show t <> "\""
+
+instance ToJSON EventSource where
+  toJSON = \case
+    AwsKafka -> String "aws:kafka"
+    SelfManagedKafka -> String "SelfManagedKafka"
+
+-- | Convenience alias: most of the time you will parse the records
+-- straight into some app-specific structure.
+type Record = Record' ByteString
+
+-- | Records from a Kafka event. This is 'Traversable', which means
+-- you can do things like parse a JSON-encoded payload:
+--
+-- @
+-- 'traverse' 'decodeStrict' :: 'FromJSON' a => Record ByteString -> Maybe (Record a)
+-- @
+data Record' a = Record {
+  topic :: !Text,
+  partition :: !Int32,
+  offset :: !Int64,
+  timestamp :: !Timestamp,
+  -- | NOTE: there can be multiple headers for a given key.
+  headers :: !(Vector Header),
+  key :: !(Maybe ByteString),
+  value :: !(Maybe a)
+} deriving (Eq, Show, Generic, Functor, Foldable, Traversable)
+
+-- | Decodes base64-encoded keys and values, where present.
+instance a ~ ByteString => FromJSON (Record' a) where
+  parseJSON = withObject "Record" $ \o -> do
+    topic <- o .: "topic"
+    partition <- o .: "partition"
+    offset <- o .: "offset"
+    timestamp <- parseTimestamp o
+    headers <- o .: "headers"
+    key <- fmap (B64.decodeLenient . TE.encodeUtf8) <$> o .:? "key"
+    value <- fmap (B64.decodeLenient . TE.encodeUtf8) <$> o .:? "value"
+    pure Record { topic, partition, offset, timestamp, headers, key, value }
+
+-- | Encodes keys and values into base64.
+instance a ~ ByteString => ToJSON (Record' a) where
```

Comment:
Is `a ~ ByteString => ToJSON (Record' a)` equivalent to `ToJSON (Record' ByteString)`, or does this approach have an advantage?